src/search/cost.cc

```diff
@@ -192,9 +194,291 @@ bool is_win_legal(inst* orig, int len1, inst* synth, int len2,
+  if (num_successful_ex == ex_set_size) {
+    auto t1 = NOW;
+    try {
+      is_equal = _vld.is_equal_to(orig->inst_list, len1, synth->inst_list, len2);
+      // if (smt_var::is_win) {
+      //   // check win prog eq check result
+      //   validator vld1;
+      //   vld1._is_win = false;
+      //   smt_var::is_win = false;
+      //   vld1.set_orig(orig->inst_list, len1);
+      //   int is_equal_expected = vld1.is_equal_to(orig->inst_list, len1, synth->inst_list, len2);
+      //   cout << "win prog check: " << is_equal << " " << is_equal_expected << endl;
+      //   if (is_equal_expected != is_equal) {
+      //     cout << "win prog check fail: " << is_equal << " " << is_equal_expected << endl;
+      //   }
+      //   smt_var::is_win = true;
+      // }
+    } catch (const string err_msg) {
+      // illegal program
+      //synth->set_error_cost(ERROR_COST_MAX);
+      //synth->set_safety_cost(ERROR_COST_MAX);
+      //return ERROR_COST_MAX;
+      //already took care of this situation for safety
+    }
+    auto t2 = NOW;
+    auto dur = DUR(t1, t2);
+    // cout << dur << endl;
+    dur_sum += dur;
+    if (dur > 50000) {
+      dur_sum_long += dur;
+      n_sum_long++;
+      // synth->print();
+    }
+  }
+  int avg_value = get_avg_value(ex_set_size);
+  total_cost = get_final_error_cost(total_cost, is_equal,
+                                    ex_set_size, num_successful_ex,
+                                    avg_value);
+  // process counterexamples
+  // If num_successful_ex < (int)_examples._exs.size(),
+  // it shows the example that synth fails in the example set is a counterexample.
+  // The counterexample generated from this synth may have already been in the examples set.
+  // Thus, only when num_successful_ex == (int)_examples._exs.size(),
+  // the counterexample generated from this synth must can be added into the example set.
+  // But it should ensure that the number of initial example set is big enough.
+  // case 1: gen_counterex_flag = (is_equal == 0);
+  // case 2: gen_counterex_flag = (is_equal == 0) && (num_successful_ex == (int)_examples._exs.size());
+  //if (((is_equal == 0) || (is_equal == ILLEGAL_CEX)) &&
+   //   (num_successful_ex == (int)_examples._exs.size())) {
+  //we already take care of is_equal < 0 for safety
+  //New condition:  
+  if((num_successful_ex == (int)_examples._exs.size()) && (is_equal == 0)){
+    _examples.insert(_vld._last_counterex);
+    _meas_new_counterex_gened = true;
+    cout << "counterexample " << _examples.size() << ":" << endl;
+    cout << _vld._last_counterex.input << endl;
+    cout << _vld._last_counterex.output << endl;
+  }
+  // in case there is overflow which makes a positive value become a negative value or
+  // total_cost > ERROR_COST_MAX
+  // if ((total_cost > ERROR_COST_MAX) || (total_cost < 0)) {
+  //   synth->set_error_cost(ERROR_COST_MAX);
+  //   return ERROR_COST_MAX;
+  // }
+
+  synth->set_error_cost(total_cost);
+  return total_cost;
```

Comment:
It seems like there is a lot of code duplication between the safety cost computation and the error cost computtion, and also between the error cost computation here (doing repair) and error cost while doing optimization.

Code duplication makes it much harder later on to understand and debug code, especially the kind of code base we have. Refactor to avoid code duplication!

Reply:
Hi Professor, yes this makes sense. So shall I try to add helper functions for any general repeating code in these 3 functions, like static safety, interpretation loop, etc ..? Or are you just referring to this particular instance (generating counter example)?